eclass/dotnet-utils.eclass

```diff
@@ -0,0 +1,216 @@
+# Copyright 1999-2023 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+# @ECLASS: dotnet-utils.eclass
+# @MAINTAINER:
+# anna@navirc.com
+# @AUTHOR:
+# Anna Figueiredo Gomes <anna@navirc.com>
+# @SUPPORTED_EAPIS: 7 8
+# @BLURB: common functions and variables for dotnet builds
+
+case ${EAPI} in
+	7|8) ;;
+	*) die "${ECLASS}: EAPI ${EAPI} unsupported." ;;
+esac
+
+if [[ ! ${_DOTNET_UTILS_ECLASS} ]]; then
+_DOTNET_UTILS_ECLASS=1
+
+inherit edo multiprocessing
+
+# @ECLASS_VARIABLE: DOTNET_SLOT
+# @REQUIRED
+# @PRE_INHERIT
+# @DESCRIPTION:
+# Allows to choose a slot for dotnet
+
+if [[ ! ${DOTNET_SLOT} ]]; then
+	die "${ECLASS}: DOTNET_SLOT not set"
+fi
+
+# Temporary, use the virtual once you can have multiple virtuals installed at once
+BDEPEND+=" || ( dev-dotnet/dotnet-sdk-bin:${DOTNET_SLOT} dev-dotnet/dotnet-sdk:${DOTNET_SLOT} )"
+
+IUSE+=" debug"
+
+# @ECLASS_VARIABLE: DOTNET_EXECUTABLE
+# @INTERNAL
+# @DESCRIPTION:
+# Sets to the valid path of a dotnet executable
+
+# @ECLASS_VARIABLE: DOTNET_OUTPUT
+# @OUTPUT_VARIABLE
+# @DESCRIPTION:
+# Path of the output directory, if unset, is set during src_compile to:
+# "${WORKDIR}"/${P}_net${DOTNET_SLOT}_${DOTNET_CONFIGURATION}/${PN}
+
+# @ECLASS_VARIABLE: DOTNET_CONFIGURATION
+# @DEFAULT_UNSET
+# @DESCRIPTION:
+# Configuration value passed to dotnet in the compile phase.
+# Defaults to Debug or Release, depending on the debug useflag.
+
+# @ECLASS_VARIABLE: DOTNET_CLI_TELEMETRY_OPTOUT
+# @OUTPUT_VARIABLE
+# @DESCRIPTION:
+# Disables telemetry on dotnet.
+export DOTNET_CLI_TELEMETRY_OPTOUT=1
+
+# @ECLASS_VARIABLE: MSBUILDDISABLENODEREUSE
+# @OUTPUT_VARIABLE
+# @DESCRIPTION:
+# Stops the dotnet node after the build is done.
+export MSBUILDDISABLENODEREUSE=1
+
+# @ECLASS_VARIABLE: DOTNET_NOLOGO
+# @OUTPUT_VARIABLE
+# @DESCRIPTION:
+# Disables the header logo when running dotnet commands.
+export DOTNET_NOLOGO=1
+
+# Needed otherwise the binaries may break
+RESTRICT+=" strip"
+
+# @ECLASS_VARIABLE: NUGETS
+# @DEFAULT_UNSET
+# @DESCRIPTION:
+# String containing all nuget packages that need to be downloaded.
```

Comment:
How do I, as a developer who knows their software uses dotnet, figure out what to put in this variable?

Reply:
What I have been doing so far is manually getting them by first unpacking the source, then:
```sh
mkdir pkg
dotnet restore --runtime linux --packages pkg # Saving the nugets to pkg
cd pkg
for i in */*; do
    echo $i | sed 's#/#-#' >> requirements
done
```
The contents of `requirements` should be usable as the value of NUGETS

Maybe a script that doesn't need to download the nugets once could be made, using the nuget repo API, but recursively solving the deps sounds quite complex.

Reply:
> dotnet restore --runtime linux --packages pkg # Saving the nugets to pkg

I was doing it with `NUGET_PACKAGES=$(pwd)/nuget_packages dotnet restore`

Reply:
```
-r|--runtime <RUNTIME_IDENTIFIER>

Specifies a runtime for the package restore. This is used to restore packages for runtimes not explicitly listed in the <RuntimeIdentifiers> tag in the .csproj file. For a list of Runtime Identifiers (RIDs), see the [RID catalog](https://learn.microsoft.com/en-us/dotnet/core/rid-catalog).
```

By what I understood, if you specify --runtime linux, it will fetch only the nugets for linux and derivate runtimes. If you don't specify the runtime, it'll try to fetch nugets for every runtime specified in the csproj. Tested with two dotnet packages:

![xivlauncher](https://user-images.githubusercontent.com/117868723/215357380-098e36c7-5fcb-4957-bb14-8d1e2ab72af5.png)

On xivlauncher, it fetched extra packages, because windows is also specified on the csproj

![csharp](https://user-images.githubusercontent.com/117868723/215357354-1ce5ef70-5ddf-4b75-9c23-22f5214933ed.png)

csharp-ls does not seem to specify a RID on it's fsproj at all, so I think it auto-detected my machine as linux-x64 and only got the packages to it, while getting more packages when specifying a specific RID. (I'm not so sure if this is the case, but seems a likely explanation so far)